book/scheduling-and-threading.md

```diff
@@ -195,34 +206,53 @@ def set_timeout(func, sec):
     t.start()
 ```
 
-Next, add a dirty bit called `needs_pipeline_update` (plus `display_scheduled`
-to avoid double-running `set_timeout` unnecessarily) to `Tab`, which means
-"rendering needs to happen, and has been scheduled, but hasn't happened yet".
- 
-Also, rename `render` to `run_rendering_pipeline`, and add a new
-`run_animation_frame` method that runs the pipeline and calls the other
-rendering pipeline stages on `Browser` via a new `raster_and_draw`
-method (which we'll implement shortly).
+Next, add three dirty bits to `Tab`:
+
+* `needs_pipeline_update`, indicating that
+the pipeline needs to be re-run.
+* `display_scheduled`, indicating that `set_timeout` was already called (this
+avoids double-running `set_timeout` unnecessarily).
```

Comment:
I'm a little confused:

1) Why it's called `display_scheduled`, if it's about `set_timeout`
2) What does this mean by double-running `set_timeout`?
3) How these flags relate to each other.

Reply:
More precisely, I think this goes back to my concern that there are a lot of flags and we haven't given the reader any way of thinking about them. Why so many flags? How do you remember how they relate? How do you choose to have more or fewer? I'm not actually sure if we want these dirty flags to be a major player in this chapter, but if we do we should talk more about managing them.

Reply:
display_scheduled avoids running stuff twice, but doesn't mean the event loop should run just now. needs_animation_frame means "please run it at the next opportunity"

display_scheduled avoids double running because if you call set_timeout twice, before the callback has run, the second-time will be a no-op.

Unfortunately I do think we need at least some of these dirty flags or the browser will not function correctly (e.g. the ones above). Other flags are optional and for performance reasons. We could push them to the end of the chapter.

Reply:
Agreed, sounds good.